Don't use intermediate file when listing contents of tar.xz file

terminalmage opened the pull request:
This instead pipes the decompressed tar data directly to ``tarfile.open()``

Thanks to @dwfreed for working with me on this at PyCon!

Resolves #41190.

terminalmage committed:
Don't use intermediate file when listing contents of tar.xz file

This instead pipes the decompressed tar data directly to tarfile.open()

salt-jenkins commented:
@terminalmage, thanks for your PR! By analyzing the history of the files in this pull request, we identified @opdude, @jfindlay and @techhat to be potential reviewers.

dwfreed commented on salt/modules/archive.py:
```diff
@@ -157,7 +158,10 @@ def _list_tar(name, cached, decompress_cmd, failhard=False):
-            with contextlib.closing(tarfile.open(cached)) as tar_archive:
+            open_kwargs = {'name': cached} \
+                if not isinstance(cached, subprocess.Popen) \
+                else {'fileobj': cached.stdout, 'mode': 'r|*'}
```
You can actually drop the * here.  The subprocess should be returning an uncompressed tar archive, so there's no need to run the decompression heuristic on it.

terminalmage replied:
Good call, Just pushed another commit to drop the asterisk.

terminalmage committed:
Remove '*' from mode

This is unnecessary, as the data being piped in is uncompressed.

dwfreed commented on salt/modules/archive.py:
```diff
@@ -168,7 +172,15 @@ def _list_tar(name, cached, decompress_cmd, failhard=False):
         except tarfile.ReadError:
-            if not failhard:
+            if failhard:
+                if fileobj:
```
This should be `if isinstance(cached, subprocess.Popen):`

terminalmage replied:
Yeah, that was leftover from earlier. I had initially added a ``fileobj`` arg to ``_list_tar()`` and then decided to remove it and instead just check if ``cached`` was a Popen object. I just didn't catch this other reference. Thanks!

terminalmage committed:
Replace reference to fileobj

This was an argument I had initially added to _list_tar() but later
removed in favor of an isinstance check. Since this was removed, this
argument no longer exists.

terminalmage committed:
Remove unused import

cachedout reviewed (approved):
Clever! I dig it.
